ledger-tool/src/main.rs

```diff
@@ -4185,55 +4189,140 @@ fn main() {
                     force_update_to_open,
                 );
+
                 match blockstore.slot_meta_iterator(0) {
                     Ok(metas) => {
-                        let all = arg_matches.is_present("all");
+                        let output_format =
+                            OutputFormat::from_matches(arg_matches, "output_format", false);
+                        let all = matches.is_present("all");
 
                         let slots: Vec<_> = metas.map(|(slot, _)| slot).collect();
-                        if slots.is_empty() {
-                            println!("Ledger is empty");
+
+                        let slot_bounds = if slots.is_empty() {
+                            SlotBounds::default()
                         } else {
-                            let first = slots.first().unwrap();
-                            let last = slots.last().unwrap_or(first);
-                            if first != last {
-                                println!(
-                                    "Ledger has data for {} slots {:?} to {:?}",
-                                    slots.len(),
-                                    first,
-                                    last
-                                );
-                                if all {
-                                    println!("Non-empty slots: {slots:?}");
+                            
+                            // Collect info about slot bounds
+                            let mut bounds = SlotBounds {
+                                slots: Some(SlotInfo {
+                                    total: slots.len(),
+                                    first: Some(*slots.first().unwrap()),
+                                    last: Some(*slots.last().unwrap()),
+                                    ..Default::default()
+                                }),
+                                ..Default::default()
+                            };
+
+                            // Consider also rooted slots, if present
+                            if let Ok(rooted) = blockstore.rooted_slot_iterator(0) {
+                                let mut first_rooted = 0;
+                                let mut last_rooted = 0;
+                                let mut total_rooted = 0;
+                                for (i, slot) in rooted.into_iter().enumerate() {
+                                    if i == 0 {
+                                        first_rooted = slot;
+                                    }
+                                    last_rooted = slot;
+                                    total_rooted += 1;
                                 }
-                            } else {
-                                println!("Ledger has data for slot {first:?}");
+
+                                let count_past_root = slots.iter()
+                                    .rev()
+                                    .take_while(|slot| *slot > &last_rooted)
```

Comment:
Not sure that the `&` is necessary here

Reply:
Yes it is needed, 'cause the closure takes a `slot` that is a `&&u64`, so the compiler suggested

Reply:
Yeah, it wasn't immediately obvious to me that within the closure `slot` would be `&&u64`, but I should have checked [the docs](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) as they explicitly state it out.

> Because the closure passed to take_while() takes a reference, and many iterators iterate over references, this leads to a possibly confusing situation, where the type of the closure is a double reference: